proposals/XXXX-spirv-resource-representation.md

```diff
@@ -0,0 +1,271 @@
+
+1.  They all contains an array of memory that maps to a storage buffer.
+2.  Other than `StructuredBuffer`, they all contains a separate counter variable
+    that is its own storage buffer.
+3.  The references to `RasterizerOrderedStructuredBuffer` are contained in
+    implicit critical regions. In SPIR-V, explicit instructions are used to
+    start and stop the critical region.
+
+This makes it impossible to create a handle type that maps directly to a SPIR-V
+type. To handle this, we will create a target type `spv.Buffer`:
+
+```
+target("spv.Buffer", ElementType, StorageClass, IsWriteable, IsROV)
+target("spv.Buffer", ElementType, StorageClass, IsWriteable, IsROV, CounterSet, CounterBinding)
+```
+
+`ElementType` is the type for the storage buffer array, and `StorageClass` is
+the storage class for the array. `IsWritable` is true of the resource an be
+written to, and `IsROV` is true if it is a rasterizer order view. If the
+resource has an associated counter variable, its set and binding can be provided
+in `CounterSet` and `CounterBinding`.
+
+In the SPIR-V backend, there will be a legalization pass that will lower the
+`spv.Buffer` type to code closer to the SPIR-V to be generated:
+
+1.  Calls to `@llvm.spv.handle.fromBinding` will be replaced by two calls. One
+    that returns a handle to the array, and another that return a handle to the
+    counter, if necessary.
+2.  Calls to `@llvm.spv.resource.getpointer` will have the handle replaced by
+    the handle of the array.
+3.  Calls to `@llvm.spv.resource.updatecounter` will be replaced by a call to
+    `@llvm.spv.resource.getpointer` with the handle of the counter followed by
+    an atomic add.
+4.  If the type of the original handle is rasterizer ordered, all uses of
+    `@llvm.spv.resource.getpointer` will be surrounded by instructions to begin
+    and end the critical region.
+
+A separate legalization pass will then move the critical region markers so that
+they follow the rules required by the SPIR-V specification. This will be the
+same as the
+[`InvocationInterlockPlacementPass`](https://github.com/KhronosGroup/SPIRV-Tools/blob/682bcd51548e670811f1d03511968bb59a1157ce/source/opt/invocation_interlock_placement_pass.h)
+pass in SPIR-V Tools.
+
+The types for the handles will be target types that represent pointers. The
+handle for the array will be
+
+```llvm-ir
+%T = type { ... } ; Fully laid out version of T.
+%T1 = type { [0 x %T] } ; The SPIR-V backend should turn the array into a runtime array.
+target("spirv.Type", target(spirv.Literal, StorageClass), %T1,
+/* OpTypePointer */32)
+```
+
+TODO: I need to check if a call to `int_spv_assign_decoration` can be added that
+will decorate the target type with the block decoration.
+
+The types for the buffers must have an
+[explicit layout](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#ExplicitLayout).
+The layout information will be obtained from the DataLayout class:
```

Comment:
This is being discussed in the constant buffer proposal. See https://github.com/llvm/wg-hlsl/pull/94#discussion_r1943104242.

Reply:
In the short term, we will go with this plan. We can revisit when we add `vk::offset`.